File: sdk/ts/src/workspaceRestore.ts
import type { Turn } from './types';

export interface WorkspaceWindowSnapshot {
    windowId: string;
    conversationId?: string | null;
    windowKey: string;
    windowTitle?: string | null;
    presentation?: string | null;
    region?: string | null;
    parentKey?: string | null;
    workspaceSharePct?: number | null;
    workspaceMinHeight?: number | null;
    inTab?: boolean;
    parameters?: Record<string, unknown>;
    windowForm?: Record<string, unknown>;
}

export interface HostedWorkspaceRestoreState {
    windows: WorkspaceWindowSnapshot[];
    selectedWindowId?: string | null;
}

function parsePayload(raw: unknown): any {
    if (!raw) return null;
    if (typeof raw === 'string') {
        try {
            return JSON.parse(raw);
        } catch {
            return null;
        }
    }
    if (typeof raw === 'object') {
        const inlineBody = (raw as any)?.inlineBody ?? (raw as any)?.InlineBody;
        if (typeof inlineBody === 'string') {
            try {
                return JSON.parse(inlineBody);
            } catch {
                return raw;
            }
        }
        return raw;
    }
    return null;
}

function isPayloadEnvelope(value: any): boolean {
    if (!value || typeof value !== 'object' || Array.isArray(value)) return false;
    const hasInlineBody = typeof value.inlineBody === 'string' || typeof value.InlineBody === 'string';
    const hasCompression = typeof value.compression === 'string' || typeof value.Compression === 'string';
    const hasDirectWorkspaceShape = 'items' in value || 'windowId' in value || 'focusedWindowId' in value;
    return (hasInlineBody || hasCompression) && !hasDirectWorkspaceShape;
}

function firstParsedPayload(...candidates: unknown[]): any {
    for (const candidate of candidates) {
        const parsed = parsePayload(candidate);
        if (isPayloadEnvelope(parsed)) continue;
        if (parsed && typeof parsed === 'object') {
            return parsed;
        }
    }
    return null;
}

function normalizeToolName(raw: unknown): string {
    return String(raw || '').trim().toLowerCase().replace(/:/g, '/');
}

function toolStepsForTurn(turn: Turn | null | undefined): any[] {
    const currentTurn = turn as any;
    if (!currentTurn || typeof currentTurn !== 'object') return [];
    const pages = Array.isArray(currentTurn?.execution?.pages) ? currentTurn.execution.pages : [];
    const result: any[] = [];
    for (const page of pages) {
        const toolSteps = Array.isArray(page?.toolSteps) ? page.toolSteps : [];
        for (const step of toolSteps) {
            result.push(step || {});
        }
    }
    return result;
}

function normalizeHostedWorkspaceWindow(raw: any): WorkspaceWindowSnapshot | null {
    if (!raw || typeof raw !== 'object') return null;
    const parentKey = String(raw.parentKey || '').trim();
    const windowId = String(raw.windowId || '').trim();
    const windowKey = String(raw.windowKey || '').trim();
    if (!windowId || !windowKey) return null;
    const parameters = raw.parameters && typeof raw.parameters === 'object'
        ? raw.parameters as Record<string, unknown>
        : {};
    const windowForm = raw.windowForm && typeof raw.windowForm === 'object'
        ? raw.windowForm as Record<string, unknown>
        : undefined;
    return {
        windowId,
        conversationId: String(raw.conversationId || '').trim() || null,
        windowKey,
        windowTitle: String(raw.windowTitle || '').trim() || windowKey,
        presentation: raw.presentation || null,
        region: raw.region || null,
        parentKey,
        workspaceSharePct: Number.isFinite(Number(raw.workspaceSharePct)) ? Number(raw.workspaceSharePct) : undefined,
        workspaceMinHeight: Number.isFinite(Number(raw.workspaceMinHeight)) ? Number(raw.workspaceMinHeight) : undefined,
        inTab: raw.inTab !== false,
        parameters,
        windowForm,
    };
}

function hostedWorkspaceWindowsFromListPayload(raw: unknown): WorkspaceWindowSnapshot[] {
    const payload = firstParsedPayload(raw);
    const items = Array.isArray(payload?.items) ? payload.items : [];
    return items
        .map((item) => normalizeHostedWorkspaceWindow(item))
        .filter((item): item is WorkspaceWindowSnapshot => !!item);
}

function isPlainObject(value: unknown): value is Record<string, unknown> {
    return !!value && typeof value === 'object' && !Array.isArray(value);
}

function cloneValue<T>(value: T): T {
    return value == null ? value : JSON.parse(JSON.stringify(value));
}

function mergeWindowFormValue(base: unknown, patch: unknown): unknown {
    if (patch === undefined) return cloneValue(base);
    if (base === undefined) return cloneValue(patch);
    if (Array.isArray(base) || Array.isArray(patch)) return cloneValue(patch);
    if (!isPlainObject(base) || !isPlainObject(patch)) return cloneValue(patch);
    const merged: Record<string, unknown> = { ...base };
    const keys = new Set([...Object.keys(base), ...Object.keys(patch)]);
    keys.forEach((key) => {
        merged[key] = mergeWindowFormValue(base[key], patch[key]);
    });
    return merged;
}

function selectedWindowIdFromToolSteps(toolSteps: any[], windows: WorkspaceWindowSnapshot[]): string {
    const windowIds = new Set(windows.map((window) => String(window.windowId || '').trim()).filter(Boolean));
    if (windowIds.size === 0) return '';
    for (let i = toolSteps.length - 1; i >= 0; i -= 1) {
        const step = toolSteps[i] || {};
        if (String(step?.status || '').trim().toLowerCase() !== 'completed') continue;
        const toolName = normalizeToolName(step?.toolName);
        if (toolName === 'ui/window/show') {
            const requestPayload = parsePayload(step?.requestPayload);
            const windowId = String(requestPayload?.windowId || '').trim();
            if (windowIds.has(windowId)) return windowId;
        }
        if (toolName === 'ui/window/list') {
            const responsePayload = firstParsedPayload(step?.responsePayload, step?.content);
            const focusedWindowId = String(responsePayload?.focusedWindowId || '').trim();
            if (windowIds.has(focusedWindowId)) return focusedWindowId;
        }
    }
    return '';
}

function hostedWorkspaceWindowsFromViewOpenStep(step: any): WorkspaceWindowSnapshot[] {
    const responsePayload = firstParsedPayload(step?.responsePayload, step?.content);
    const requestPayload = firstParsedPayload(step?.requestPayload);
    const items = Array.isArray(responsePayload?.items) ? responsePayload.items : [];
    if (items.length > 0) {
        return items
            .map((item) => normalizeHostedWorkspaceWindow(item))
            .filter((item): item is WorkspaceWindowSnapshot => !!item);
    }
    const normalized = normalizeHostedWorkspaceWindow({
        windowId: String(responsePayload?.windowId || '').trim(),
        conversationId: String(responsePayload?.conversationId || '').trim() || null,
        windowKey: String(responsePayload?.windowKey || requestPayload?.id || requestPayload?.windowKey || '').trim(),
        windowTitle: String(responsePayload?.windowTitle || '').trim(),
        presentation: String(responsePayload?.presentation || '').trim(),
        region: String(responsePayload?.region || '').trim(),
        parentKey: String(responsePayload?.parentKey || '').trim(),
        workspaceSharePct: Number.isFinite(Number(responsePayload?.workspaceSharePct))
            ? Number(responsePayload?.workspaceSharePct)
            : undefined,
        workspaceMinHeight: Number.isFinite(Number(responsePayload?.workspaceMinHeight))
            ? Number(responsePayload?.workspaceMinHeight)
            : undefined,
        inTab: responsePayload?.inTab !== false,
        parameters: requestPayload?.parameters && typeof requestPayload.parameters === 'object' ? requestPayload.parameters : {},
        windowForm: responsePayload?.windowForm && typeof responsePayload.windowForm === 'object'
            ? responsePayload.windowForm
            : undefined,
    });
    return normalized ? [normalized] : [];
}

function targetHostedWorkspaceWindows(windows: WorkspaceWindowSnapshot[], requestPayload: any): WorkspaceWindowSnapshot[] {
    const targetWindowId = String(requestPayload?.windowId || '').trim();
    if (targetWindowId) {
        return windows.filter((window) => String(window?.windowId || '').trim() === targetWindowId);
    }
    const targetWindowKey = String(requestPayload?.windowKey || '').trim();
    if (!targetWindowKey) {
        return [];
    }
    const matches = windows.filter((window) => String(window?.windowKey || '').trim() === targetWindowKey);
    return matches.length === 1 ? matches : [];
}

function applySetFormDataSteps(toolSteps: any[], windows: WorkspaceWindowSnapshot[], baseIndex: number): WorkspaceWindowSnapshot[] {
    const resolved = windows.map((window) => ({
        ...window,
        ...(window.windowForm && isPlainObject(window.windowForm)
            ? { windowForm: cloneValue(window.windowForm) as Record<string, unknown> }
            : {}),
    }));
    for (let index = Math.max(baseIndex + 1, 0); index < toolSteps.length; index += 1) {
        const step = toolSteps[index] || {};
        if (String(step?.status || '').trim().toLowerCase() !== 'completed') continue;
        const toolName = normalizeToolName(step?.toolName);
        if (toolName !== 'ui/window/setformdata') continue;
        const requestPayload = firstParsedPayload(step?.requestPayload);
        const responsePayload = firstParsedPayload(step?.responsePayload, step?.content);
        const authoritativeWindowForm = responsePayload?.windowForm;
        const values = isPlainObject(authoritativeWindowForm)
            ? authoritativeWindowForm
            : requestPayload?.values;
        if (!isPlainObject(values)) continue;
        const replace = isPlainObject(authoritativeWindowForm) || requestPayload?.replace === true;
        const targetPayload = responsePayload?.windowId || responsePayload?.windowKey
            ? responsePayload
            : requestPayload;
        const targets = targetHostedWorkspaceWindows(resolved, targetPayload);
        if (targets.length === 0) continue;
        targets.forEach((window) => {
            window.windowForm = replace
                ? cloneValue(values) as Record<string, unknown>
                : mergeWindowFormValue(window.windowForm, values) as Record<string, unknown>;
        });
    }
    return resolved;
}

export function deriveHostedWorkspaceRestoreStateFromTranscriptTurns(turns: Turn[] = []): HostedWorkspaceRestoreState | null {
    const list = Array.isArray(turns) ? turns : [];
    const toolSteps = toolStepsForTurn(list[list.length - 1]);
    if (toolSteps.length === 0) return null;
    for (let i = toolSteps.length - 1; i >= 0; i -= 1) {
        const step = toolSteps[i] || {};
        if (String(step?.status || '').trim().toLowerCase() !== 'completed') continue;
        const toolName = normalizeToolName(step?.toolName);
        if (toolName === 'ui/window/list') {
            const windows = applySetFormDataSteps(
                toolSteps,
                hostedWorkspaceWindowsFromListPayload(firstParsedPayload(step?.responsePayload, step?.content)),
                i,
            );
            if (windows.length === 0) continue;
            return {
                windows,
                selectedWindowId: selectedWindowIdFromToolSteps(toolSteps, windows) || null,
            };
        }
        if (toolName === 'ui/view/open' || toolName === 'ui/window/open') {
            const windows = applySetFormDataSteps(toolSteps, hostedWorkspaceWindowsFromViewOpenStep(step), i);
            if (windows.length === 0) continue;
            const responsePayload = firstParsedPayload(step?.responsePayload, step?.content);
            const selectedWindowId = String(responsePayload?.selectedWindowId || '').trim()
                || String(windows[windows.length - 1]?.windowId || '').trim();
            return {
                windows,
                selectedWindowId: selectedWindowId || null,
            };
        }
    }
    return null;
}

File: sdk/ios/Sources/AgentlySDK/WorkspaceRestore.swift
import Foundation

public func deriveHostedWorkspaceRestoreState(
    from response: ConversationStateResponse
) -> HostedWorkspaceRestoreState? {
    let toolSteps = response.conversation?.turns.flatMap { turn in
        turn.execution?.pages.flatMap(\.toolSteps).map(HostedWorkspaceToolStep.init) ?? []
    } ?? []
    guard !toolSteps.isEmpty else { return nil }
    return deriveHostedWorkspaceRestoreState(from: toolSteps)
}

public func deriveHostedWorkspaceRestoreState(
    from response: ConversationStateResponse?,
    streamSnapshot: ConversationStreamSnapshot?
) -> HostedWorkspaceRestoreState? {
    if let streamSnapshot,
       streamSnapshot.activeTurnID?.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty == false {
        return deriveHostedWorkspaceRestoreState(from: streamSnapshot)
    }
    if let streamSnapshot, !streamSnapshot.liveExecutionGroupsByID.isEmpty {
        return nil
    }
    return response.flatMap { deriveHostedWorkspaceRestoreState(from: $0) }
}

public func deriveHostedWorkspaceRestoreState(
    from snapshot: ConversationStreamSnapshot
) -> HostedWorkspaceRestoreState? {
    let targetTurnID = snapshot.activeTurnID?.trimmedNonEmpty ?? ""
    guard !targetTurnID.isEmpty else {
        return nil
    }
    let groups = liveHostedWorkspaceGroups(snapshot, targetTurnID: targetTurnID)
    guard !groups.isEmpty else {
        return nil
    }
    let toolSteps = groups.flatMap { group in
        group.toolSteps.map(HostedWorkspaceToolStep.init)
    }
    return deriveHostedWorkspaceRestoreState(from: toolSteps)
}

private struct HostedWorkspaceToolStep {
    let toolName: String?
    let status: String?
    let content: String?
    let requestPayload: JSONValue?
    let responsePayload: JSONValue?

    init(_ step: ToolStepState) {
        self.toolName = step.toolName
        self.status = step.status
        self.content = step.content
        self.requestPayload = step.requestPayload
        self.responsePayload = step.responsePayload
    }

    init(_ step: LiveToolStepState) {
        self.toolName = step.toolName
        self.status = step.status
        self.content = nil
        self.requestPayload = step.requestPayload
        self.responsePayload = step.responsePayload
    }
}

private func deriveHostedWorkspaceRestoreState(
    from toolSteps: [HostedWorkspaceToolStep]
) -> HostedWorkspaceRestoreState? {
    let completedToolSteps = toolSteps.filter {
        ($0.status ?? "").trimmingCharacters(in: .whitespacesAndNewlines).lowercased() == "completed"
    }
    for index in completedToolSteps.indices.reversed() {
        let step = completedToolSteps[index]
        let allCompletedToolSteps = completedToolSteps[...]
        let toolName = normalizeHostedWorkspaceToolName(step.toolName)
        if toolName == "ui/window/list" {
            let windows = applyWindowFormDataPatches(
                hostedWorkspaceWindowsFromListPayload(
                    firstParsedPayload(step.responsePayload, step.content)
                ),
                allCompletedToolSteps
            )
            if !windows.isEmpty {
                let selectedWindowID = selectedWindowIDFromToolSteps(completedToolSteps, windows)
                return HostedWorkspaceRestoreState(
                    windows: windows,
                    selectedWindowId: selectedWindowID.isEmpty ? windows.last?.windowId : selectedWindowID
                )
            }
        }
        if toolName == "ui/view/open" || toolName == "ui/window/open" {
            let windows = applyWindowFormDataPatches(
                hostedWorkspaceWindowsFromViewOpenStep(step),
                allCompletedToolSteps
            )
            if !windows.isEmpty {
                let responsePayload = firstParsedPayload(step.responsePayload, step.content)
                let selectedWindowID = responsePayload?.objectValue?["selectedWindowId"]?.stringValue
                    ?? windows.last?.windowId
                return HostedWorkspaceRestoreState(
                    windows: windows,
                    selectedWindowId: selectedWindowID?.isEmpty == false ? selectedWindowID : nil
                )
            }
        }
    }
    return nil
}

private func applyWindowFormDataPatches(
    _ windows: [WorkspaceWindowSnapshot],
    _ toolSteps: ArraySlice<HostedWorkspaceToolStep>
) -> [WorkspaceWindowSnapshot] {
    guard !windows.isEmpty, !toolSteps.isEmpty else {
        return windows
    }
    var patchedWindows = windows
    for step in toolSteps {
        if normalizeHostedWorkspaceToolName(step.toolName) == "ui/window/get",
           let snapshot = hostedWorkspaceWindowFromGetStep(step) {
            let targets = hostedWorkspaceFormPatchTargets(
                windows: patchedWindows,
                windowID: snapshot.windowId,
                windowKey: snapshot.windowKey
            )
            if !targets.isEmpty {
                patchedWindows = patchedWindows.map { window in
                    targets.contains(window.windowId) ? snapshot : window
                }
            }
            continue
        }
        guard normalizeHostedWorkspaceToolName(step.toolName) == "ui/window/setformdata" else {
            continue
        }
        let requestPayload = firstParsedPayload(step.requestPayload, nil)?.objectValue ?? [:]
        let responsePayload = firstParsedPayload(step.responsePayload, step.content)?.objectValue ?? [:]
        let windowID = responsePayload["windowId"]?.stringValue
            ?? requestPayload["windowId"]?.stringValue
            ?? ""
        let windowKey = responsePayload["windowKey"]?.stringValue
            ?? requestPayload["windowKey"]?.stringValue
            ?? ""
        let authoritativeWindowForm = responsePayload["windowForm"]?.objectValue
        guard let nextValues = authoritativeWindowForm
            ?? requestPayload["values"]?.objectValue
            ?? requestPayload["parameters"]?.objectValue else {
            continue
        }
        let targetWindowIDs = hostedWorkspaceFormPatchTargets(
            windows: patchedWindows,
            windowID: windowID,
            windowKey: windowKey
        )
        guard !targetWindowIDs.isEmpty else {
            continue
        }
        let replace = requestPayload["replace"]?.boolValue == true
        patchedWindows = patchedWindows.map { window in
            guard targetWindowIDs.contains(window.windowId) else {
                return window
            }
            let nextWindowForm = authoritativeWindowForm != nil || replace
                ? nextValues
                : mergeJSONObjects(
                    base: invalidateDerivedReportBuilderState(
                        current: window.windowForm ?? [:],
                        patch: nextValues
                    ),
                    override: nextValues
                )
            return replacingWindowForm(window, windowForm: nextWindowForm)
        }
    }
    return patchedWindows
}

private func invalidateDerivedReportBuilderState(
    current: [String: JSONValue],
    patch: [String: JSONValue]
) -> [String: JSONValue] {
    guard patch["reportDefinition"]?.objectValue != nil else { return current }
    return current.filter { !$0.key.hasPrefix("reportBuilder:") }
}

private func hostedWorkspaceWindowFromGetStep(
    _ step: HostedWorkspaceToolStep
) -> WorkspaceWindowSnapshot? {
    guard let payload = firstParsedPayload(step.responsePayload, step.content)?.objectValue else {
        return nil
    }
    return normalizeHostedWorkspaceWindow(payload["window"]?.objectValue ?? payload)
}

private func hostedWorkspaceFormPatchTargets(
    windows: [WorkspaceWindowSnapshot],
    windowID: String,
    windowKey: String
) -> Set<String> {
    let normalizedWindowID = windowID.trimmingCharacters(in: .whitespacesAndNewlines)
    if !normalizedWindowID.isEmpty {
        return Set(windows.filter { $0.windowId == normalizedWindowID }.map(\.windowId))
    }
    let normalizedWindowKey = windowKey.trimmingCharacters(in: .whitespacesAndNewlines)
    guard !normalizedWindowKey.isEmpty else {
        return []
    }
    let matches = windows.filter { $0.windowKey == normalizedWindowKey }
    return matches.count == 1 ? Set(matches.map(\.windowId)) : []
}

private func replacingWindowForm(
    _ window: WorkspaceWindowSnapshot,
    windowForm: [String: JSONValue]
) -> WorkspaceWindowSnapshot {
    WorkspaceWindowSnapshot(
        windowId: window.windowId,
        conversationId: window.conversationId,
        windowKey: window.windowKey,
        windowTitle: window.windowTitle,
        presentation: window.presentation,
        region: window.region,
        parentKey: window.parentKey,
        workspaceSharePct: window.workspaceSharePct,
        workspaceMinHeight: window.workspaceMinHeight,
        inTab: window.inTab,
        parameters: window.parameters,
        windowForm: windowForm
    )
}

private func mergeJSONObjects(
    base: [String: JSONValue],
    override: [String: JSONValue]
) -> [String: JSONValue] {
    var merged = base
    for (key, value) in override {
        if case .object(let currentObject)? = merged[key],
           case .object(let overrideObject) = value {
            merged[key] = .object(mergeJSONObjects(base: currentObject, override: overrideObject))
        } else {
            merged[key] = value
        }
    }
    return merged
}

private func liveHostedWorkspaceGroups(
    _ snapshot: ConversationStreamSnapshot,
    targetTurnID: String
) -> [LiveExecutionGroup] {
    snapshot.liveExecutionGroupsByID.values
        .filter { group in
            group.turnID == targetTurnID
        }
        .sorted {
            let lhsSequence = $0.sequence ?? Int.max
            let rhsSequence = $1.sequence ?? Int.max
            if lhsSequence != rhsSequence { return lhsSequence < rhsSequence }
            let lhsIteration = $0.iteration ?? Int.max
            let rhsIteration = $1.iteration ?? Int.max
            if lhsIteration != rhsIteration { return lhsIteration < rhsIteration }
            return $0.pageID < $1.pageID
        }
}

private func hostedWorkspaceWindowsFromListPayload(_ raw: JSONValue?) -> [WorkspaceWindowSnapshot] {
    guard let payload = raw,
          let items = payload.objectValue?["items"]?.arrayValue else {
        return []
    }
    return items.compactMap { normalizeHostedWorkspaceWindow($0.objectValue) }
}

private func hostedWorkspaceWindowsFromViewOpenStep(_ step: HostedWorkspaceToolStep) -> [WorkspaceWindowSnapshot] {
    let responsePayload = firstParsedPayload(step.responsePayload, step.content)
    if let items = responsePayload?.objectValue?["items"]?.arrayValue, !items.isEmpty {
        return items.compactMap { normalizeHostedWorkspaceWindow($0.objectValue) }
    }
    let requestPayload = firstParsedPayload(step.requestPayload, nil)
    var merged: [String: JSONValue] = [
        "windowId": responsePayload?.objectValue?["windowId"] ?? .string(""),
        "conversationId": responsePayload?.objectValue?["conversationId"] ?? .string(""),
        "windowKey": responsePayload?.objectValue?["windowKey"]
            ?? requestPayload?.objectValue?["id"]
            ?? requestPayload?.objectValue?["windowKey"]
            ?? .string(""),
        "windowTitle": responsePayload?.objectValue?["windowTitle"] ?? .string(""),
        "presentation": responsePayload?.objectValue?["presentation"] ?? .string(""),
        "region": responsePayload?.objectValue?["region"] ?? .string(""),
        "parentKey": responsePayload?.objectValue?["parentKey"] ?? .string(""),
        "parameters": requestPayload?.objectValue?["parameters"] ?? .object([:])
    ]
    if let share = responsePayload?.objectValue?["workspaceSharePct"] {
        merged["workspaceSharePct"] = share
    }
    if let minHeight = responsePayload?.objectValue?["workspaceMinHeight"] {
        merged["workspaceMinHeight"] = minHeight
    }
    if let inTab = responsePayload?.objectValue?["inTab"] {
        merged["inTab"] = inTab
    }
    if let windowForm = responsePayload?.objectValue?["windowForm"] {
        merged["windowForm"] = windowForm
    }
    return [normalizeHostedWorkspaceWindow(merged)].compactMap { $0 }
}

private func normalizeHostedWorkspaceWindow(_ raw: [String: JSONValue]?) -> WorkspaceWindowSnapshot? {
    guard let raw else { return nil }
    let parentKey = raw["parentKey"]?.stringValue ?? ""
    let windowID = raw["windowId"]?.stringValue ?? ""
    let windowKey = raw["windowKey"]?.stringValue ?? ""
    guard !windowID.isEmpty, !windowKey.isEmpty else { return nil }
    var windowForm = raw["windowForm"]?.objectValue ?? [:]
    if let metadata = raw["metadata"] {
        windowForm["__agentlyWindowMetadata"] = metadata
    }
    return WorkspaceWindowSnapshot(
        windowId: windowID,
        conversationId: raw["conversationId"]?.stringValue?.nonEmpty,
        windowKey: windowKey,
        windowTitle: raw["windowTitle"]?.stringValue?.nonEmpty ?? windowKey,
        presentation: raw["presentation"]?.stringValue?.nonEmpty,
        region: raw["region"]?.stringValue?.nonEmpty,
        parentKey: parentKey,
        workspaceSharePct: raw["workspaceSharePct"]?.intValue,
        workspaceMinHeight: raw["workspaceMinHeight"]?.intValue,
        inTab: raw["inTab"]?.boolValue ?? true,
        parameters: raw["parameters"]?.objectValue,
        windowForm: windowForm.isEmpty ? nil : windowForm
    )
}

private func selectedWindowIDFromToolSteps(
    _ toolSteps: [HostedWorkspaceToolStep],
    _ windows: [WorkspaceWindowSnapshot]
) -> String {
    let windowIDs = Set(windows.map(\.windowId))
    for step in toolSteps.reversed() {
        let toolName = normalizeHostedWorkspaceToolName(step.toolName)
        if toolName == "ui/window/show" {
            let requestPayload = firstParsedPayload(step.requestPayload, nil)
            let windowID = requestPayload?.objectValue?["windowId"]?.stringValue ?? ""
            if windowIDs.contains(windowID) {
                return windowID
            }
        }
        if toolName == "ui/window/list" {
            let responsePayload = firstParsedPayload(step.responsePayload, step.content)
            let focusedWindowID = responsePayload?.objectValue?["focusedWindowId"]?.stringValue ?? ""
            if windowIDs.contains(focusedWindowID) {
                return focusedWindowID
            }
        }
    }
    return ""
}

private func firstParsedPayload(_ raw: JSONValue?, _ rawText: String?) -> JSONValue? {
    var candidates: [JSONValue] = []
    if let raw {
        candidates.append(raw)
    }
    if let rawText,
       !rawText.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
        candidates.append(.string(rawText))
    }
    for candidate in candidates {
        guard let parsed = parsePayload(candidate) else {
            continue
        }
        if isPayloadEnvelope(parsed) {
            continue
        }
        return parsed
    }
    return nil
}

private func parsePayload(_ raw: JSONValue) -> JSONValue? {
    switch raw {
    case .string(let value):
        let trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty, let data = trimmed.data(using: .utf8) else { return nil }
        return try? JSONDecoder.agently().decode(JSONValue.self, from: data)
    case .object(let object):
        if let inlineBody = object["inlineBody"]?.stringValue ?? object["InlineBody"]?.stringValue,
           !inlineBody.isEmpty,
           let data = inlineBody.data(using: .utf8),
           let decoded = try? JSONDecoder.agently().decode(JSONValue.self, from: data) {
            return decoded
        }
        return raw
    default:
        return raw
    }
}

private func isPayloadEnvelope(_ value: JSONValue) -> Bool {
    guard let object = value.objectValue else {
        return false
    }
    let hasInlineBody = object["inlineBody"]?.stringValue != nil || object["InlineBody"]?.stringValue != nil
    let hasCompression = object["compression"]?.stringValue != nil || object["Compression"]?.stringValue != nil
    let hasDirectWorkspaceShape = object["items"] != nil || object["windowId"] != nil || object["focusedWindowId"] != nil
    return (hasInlineBody || hasCompression) && !hasDirectWorkspaceShape
}

private func normalizeHostedWorkspaceToolName(_ raw: String?) -> String {
    String(raw ?? "")
        .trimmingCharacters(in: .whitespacesAndNewlines)
        .lowercased()
        .replacingOccurrences(of: ":", with: "/")
        .replacingOccurrences(of: ".", with: "/")
}

private extension JSONValue {
    var stringValue: String? {
        if case .string(let value) = self {
            return value.trimmingCharacters(in: .whitespacesAndNewlines)
        }
        return nil
    }

    var objectValue: [String: JSONValue]? {
        if case .object(let value) = self {
            return value
        }
        return nil
    }

    var arrayValue: [JSONValue]? {
        if case .array(let value) = self {
            return value
        }
        return nil
    }

    var intValue: Int? {
        switch self {
        case .number(let value):
            return Int(value)
        case .string(let value):
            return Int(value.trimmingCharacters(in: .whitespacesAndNewlines))
        default:
            return nil
        }
    }

    var boolValue: Bool? {
        switch self {
        case .bool(let value):
            return value
        case .string(let value):
            switch value.trimmingCharacters(in: .whitespacesAndNewlines).lowercased() {
            case "true":
                return true
            case "false":
                return false
            default:
                return nil
            }
        default:
            return nil
        }
    }
}

private extension String {
    var trimmedNonEmpty: String? {
        let trimmed = trimmingCharacters(in: .whitespacesAndNewlines)
        return trimmed.isEmpty ? nil : trimmed
    }

    var nonEmpty: String? {
        isEmpty ? nil : self
    }
}
